### DISCUSSION TITLE
Plugin to import/export/save as... a different format or a hosted service

### DISCUSSION THREAD

**Discussion by RupertBarrow** [freeplane/freeplane#1287] [Ideas]
Hi,
I dream of importing/exporting or even "saving as..." from Freeplane to Asana.
Is there a plugin model related to the data-loading and saving functions which could be overloaded to load from or save to another format ?

Adding a new import source could be a good start, as well as export. But it would be nice to have "open from Asana" and "save to Asana".

**Comment by macmarrum** [comment]
I know this page: https://docs.freeplane.org/coding/Export_Using_XSLT.html

**Comment by macmarrum** [comment]
In addition to changing Freeplane (core), it's possible to create an add-on (Groovy script/library).

Such an add-on/script could create a mind map from an external file as well as save any data from a mind map to an external file, in effect providing new import/export-format support.

### Export

Here's how an export to Markdown could be done, using my library [Export.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/Export.groovy)
1. Save `Export.groovy` to `<user-dir>/lib/` and restart Freeplane
2. Open or create a mind map with several levels, and `Format->Automatic level styles->for non leaf nodes` enabled
3. Execute the following script in Freeplane (e.g. `Tools->Edit script…`)
```groovy
import io.github.macmarrum.freeplane.Export
import javax.swing.*

final fileChooser = new JFileChooser()
fileChooser.setMultiSelectionEnabled(false)
final returnVal = fileChooser.showOpenDialog()
if (returnVal != JFileChooser.APPROVE_OPTION) {
    return
}
def file = fileChooser.getSelectedFile()

Export.exportMarkdownLevelStyles(file, node)
```

### Import
Here's my library containing a method to import a JSON representation of a mind map: [Import.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/lib/Import.groovy)

And a sample file: [freeplane1.json](https://gist.github.com/macmarrum/32913ec5f0500f7242480ae496f25f3a)

1. Save `freeplane1.json` to your disk
2. Save `Import.groovy` to `<user-dir>/lib/` and restart Freeplane
3. Execute the script in Freeplane (e.g. `Tools->Edit script…`)
```groovy
import io.github.macmarrum.freeplane.Import
import javax.swing.*

final fileChooser = new JFileChooser()
fileChooser.setMultiSelectionEnabled(false)
final returnVal = fileChooser.showOpenDialog()
if (returnVal != JFileChooser.APPROVE_OPTION) {
    return
}
def file = fileChooser.getSelectedFile()
if (!file.exists()) {
    return
}

Import.importJson(file)
```

**Reply by macmarrum** [reply]
Scripting API: https://docs.freeplane.org/api/

**Comment by RupertBarrow** [comment]
Wow, thanks for all these answers !
I now have plenty of examples to look at to try to do what I was asking for ... 😅
